flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/rules/dataset/DataSetJoinRule.scala

```diff
@@ -39,18 +49,80 @@ class DataSetJoinRule
     val convLeft: RelNode = RelOptRule.convert(join.getInput(0), DataSetConvention.INSTANCE)
     val convRight: RelNode = RelOptRule.convert(join.getInput(1), DataSetConvention.INSTANCE)
 
-    new DataSetJoin(
-      rel.getCluster,
-      traitSet,
-      convLeft,
-      convRight,
-      rel.getRowType,
-      join.toString,
-      Array[Int](),
-      Array[Int](),
-      JoinType.INNER,
-      null,
-      null)
+    // get the equality keys
+    val joinInfo = join.analyzeCondition
+    val keyPairs = joinInfo.pairs
+
+    if (keyPairs.isEmpty) { // if no equality keys => not supported
+      throw new TableException("Joins should have at least one equality condition")
+    }
+    else { // at least one equality expression => generate a join function
+      val conditionType = join.getCondition.getType
+      val func = getJoinFunction(join, joinInfo)
+      val leftKeys = ArrayBuffer.empty[Int]
+      val rightKeys = ArrayBuffer.empty[Int]
+
+      keyPairs.foreach(pair => {
+        leftKeys.add(pair.source)
+        rightKeys.add(pair.target)}
+      )
+
+      new DataSetJoin(
+        rel.getCluster,
+        traitSet,
+        convLeft,
+        convRight,
+        rel.getRowType,
+        join.toString,
+        leftKeys.toArray,
+        rightKeys.toArray,
+        JoinType.INNER,
+        null,
+        func)
+    }
+  }
+
+  def getJoinFunction(join: FlinkJoin, joinInfo: JoinInfo):
+      ((TableConfig, TypeInformation[Any], TypeInformation[Any], TypeInformation[Any]) =>
+      FlatJoinFunction[Any, Any, Any]) = {
+
+    if (joinInfo.isEqui) {
+      // only equality condition => no join function necessary
+      null
```

Comment:
In general, `null` is not very welcome in Scala. Could you return a `FlatJoinFunction` (containing only a `ConverterResultExpression`) here too? We can then get rid of the `Tuple2RowMapper` and support any type as output type of the join operation.